src/lexer/decodeHTML.ts

```diff
@@ -0,0 +1,2185 @@
+  vartriangleright: '\u22B3',
+  vcy: '\u0432',
+  vdash: '\u22A2',
+  vee: '\u2228',
+  veebar: '\u22BB',
+  veeeq: '\u225A',
+  vellip: '\u22EE',
+  verbar: '\u007C',
+  vert: '\u007C',
+  vfr: '\uD835\uDD33',
+  vltri: '\u22B2',
+  vnsub: '\u2282\u20D2',
+  vnsup: '\u2283\u20D2',
+  vopf: '\uD835\uDD67',
+  vprop: '\u221D',
+  vrtri: '\u22B3',
+  vscr: '\uD835\uDCCB',
+  vsubnE: '\u2ACB\uFE00',
+  vsubne: '\u228A\uFE00',
+  vsupnE: '\u2ACC\uFE00',
+  vsupne: '\u228B\uFE00',
+  vzigzag: '\u299A',
+  wcirc: '\u0175',
+  wedbar: '\u2A5F',
+  wedge: '\u2227',
+  wedgeq: '\u2259',
+  weierp: '\u2118',
+  wfr: '\uD835\uDD34',
+  wopf: '\uD835\uDD68',
+  wp: '\u2118',
+  wr: '\u2240',
+  wreath: '\u2240',
+  wscr: '\uD835\uDCCC',
+  xcap: '\u22C2',
+  xcirc: '\u25EF',
+  xcup: '\u22C3',
+  xdtri: '\u25BD',
+  xfr: '\uD835\uDD35',
+  xhArr: '\u27FA',
+  xharr: '\u27F7',
+  xi: '\u03BE',
+  xlArr: '\u27F8',
+  xlarr: '\u27F5',
+  xmap: '\u27FC',
+  xnis: '\u22FB',
+  xodot: '\u2A00',
+  xopf: '\uD835\uDD69',
+  xoplus: '\u2A01',
+  xotime: '\u2A02',
+  xrArr: '\u27F9',
+  xrarr: '\u27F6',
+  xscr: '\uD835\uDCCD',
+  xsqcup: '\u2A06',
+  xuplus: '\u2A04',
+  xutri: '\u25B3',
+  xvee: '\u22C1',
+  xwedge: '\u22C0',
+  yacute: '\u00FD',
+  yacy: '\u044F',
+  ycirc: '\u0177',
+  ycy: '\u044B',
+  yen: '\u00A5',
+  yfr: '\uD835\uDD36',
+  yicy: '\u0457',
+  yopf: '\uD835\uDD6A',
+  yscr: '\uD835\uDCCE',
+  yucy: '\u044E',
+  yuml: '\u00FF',
+  zacute: '\u017A',
+  zcaron: '\u017E',
+  zcy: '\u0437',
+  zdot: '\u017C',
+  zeetrf: '\u2128',
+  zeta: '\u03B6',
+  zfr: '\uD835\uDD37',
+  zhcy: '\u0436',
+  zigrarr: '\u21DD',
+  zopf: '\uD835\uDD6B',
+  zscr: '\uD835\uDCCF',
+  zwj: '\u200D',
+  zwnj: '\u200C'
+};
+
+const decodeMap: { [Identifier: string]: number } = {
+  '0': 65533,
+  '128': 8364,
+  '130': 8218,
+  '131': 402,
+  '132': 8222,
+  '133': 8230,
+  '134': 8224,
+  '135': 8225,
+  '136': 710,
+  '137': 8240,
+  '138': 352,
+  '139': 8249,
+  '140': 338,
+  '142': 381,
+  '145': 8216,
+  '146': 8217,
+  '147': 8220,
+  '148': 8221,
+  '149': 8226,
+  '150': 8211,
+  '151': 8212,
+  '152': 732,
+  '153': 8482,
+  '154': 353,
+  '155': 8250,
+  '156': 339,
+  '158': 382,
+  '159': 376
+};
+
+// Used code from https://github.com/fb55/entities (BSD license)
+export function decodeHTMLStrict(text: string): string {
+  return text.replace(/&(?:[a-zA-Z]+|#[xX][\da-fA-F]+|#\d+);/g, (key) => {
+    if (key.charAt(1) === '#') {
+      const secondChar = key.charAt(2);
+      if (secondChar === 'X' || secondChar === 'x') {
+        return decodeCodePoint(parseInt(key.substr(3), 16));
+      }
+      return decodeCodePoint(parseInt(key.substr(2), 10));
+    }
+    return entities[key.slice(1, -1)] || key;
+  });
+}
+
+function decodeCodePoint(codePoint: number): string {
+  if ((codePoint >= 0xd800 && codePoint <= 0xdfff) || codePoint > 0x10ffff) {
+    return '\uFFFD';
+  }
+
+  if (codePoint in decodeMap) {
+    codePoint = decodeMap[codePoint];
```

Comment:
I have not found the spec for this special codePoint mapping. Basically copied form fb55/entities implementation now.

Reply:
eh wow. That code consist of 2k lines? Can you check how Acorn JSX plugin does this?

Reply:
acron-jsx seems only uses part of the entities map https://github.com/acornjs/acorn-jsx/blob/master/xhtml.js

I checked few of their missed ones, React app still support them, obviously acron-jsx deliberated skipped non-popular entities.

Reply:
acron-jsx also didn't patch number code point like fb55/entities (This line I noted in this thread).
The decodeMap is a small object to translate few special code points, I have not found any spec for it.

I tested in React app, the decodeMap patch is necessary to match real app behaviour.